src/metabase/server/middleware/security.clj

```diff
@@ -124,11 +124,52 @@
           "Content-Security-Policy"
           #(format "%s frame-ancestors %s;" % (if allow-iframes? "*" (or (embedding-app-origin) "'none'")))))
 
+(defn- parse-url [url]
+  (let [pattern #"^(?:(https?)://)?([^:/]+)(?::(\d+|\*))?$"
+        matches (re-matches pattern url)]
+    (if matches
+      (let [[_ protocol domain port] matches]
+        {:protocol protocol
+         :domain domain
+         :port port})
+      (throw (IllegalArgumentException. "Invalid URL")))))
+
+(defn approved-origin?
+  "Returns true if `origin` should be allowed for CORS based on the `approved-origins`"
+  [raw-origin approved-origins]
+  (if (or (empty? raw-origin) (empty? approved-origins))
+    false
+    (let [approved-list (str/split approved-origins #" ")
+          origin        (parse-url raw-origin)]
+      (boolean (some (fn [approved-origin-raw]
+                       (or
+                        (= approved-origin-raw "*")
+                        (let [approved-origin (parse-url approved-origin-raw)]
+                          (and
+                        ;;  domain check
+                           (if (str/starts-with? (:domain approved-origin) "*.")
+                             (str/ends-with? (:domain origin) (str/replace-first (:domain approved-origin) "*." "."))
+                             (= (:domain origin) (:domain approved-origin)))
+                        ;;  protocol check
+                           (or (nil? (:protocol approved-origin))
+                               (= (:protocol origin) (:protocol approved-origin)))
+                        ;;  port check
+                           (or
+                            (= (:port approved-origin) "*")
+                            (= (:port origin) (:port approved-origin)))))))
```

Comment:
The depth of this conditional makes it a bit harder to parse what is going on. What do you think of extracting (esspecially) the domain check into a function and writing some tests for that independently?